test/hotspot/jtreg/compiler/lib/compile_framework/CompileFramework.java

```diff
@@ -0,0 +1,303 @@
+        command.add("org.openjdk.asmtools.jasm.Main");
+        command.add("-d");
+        command.add(classesDir.toString());
+        for (Path path : paths) {
+            command.add(path.toAbsolutePath().toString());
+        }
+
+        executeCompileCommand(command);
+    }
+
+    private static String[] getClassPaths() {
+        String separator = File.pathSeparator;
+        return System.getProperty("java.class.path").split(separator);
+    }
+
+    private static String getAsmToolsPath() {
+        for (String path : getClassPaths()) {
+            if (path.endsWith("jtreg.jar")) {
+                File jtreg = new File(path);
+                File dir = jtreg.getAbsoluteFile().getParentFile();
+                File asmtools = new File(dir, "asmtools.jar");
+                if (!asmtools.exists()) {
+                    throw new InternalCompileFrameworkException("Found jtreg.jar in classpath, but could not find asmtools.jar");
+                }
+                return asmtools.getAbsolutePath();
+            }
+        }
+        throw new InternalCompileFrameworkException("Could not find asmtools because could not find jtreg.jar in classpath");
+    }
+
+    private void compileJavaSources(List<SourceCode> javaSources) {
+        if (javaSources.size() == 0) {
+            println("No java sources to compile.");
+            return;
+        }
+        println("Compiling Java sources: " + javaSources.size());
+
+        List<Path> javaFilePaths = writeSourcesToFile(javaSources);
+        compileJavaFiles(javaFilePaths);
+        println("Java sources compiled.");
+    }
+
+    private void compileJavaFiles(List<Path> paths) {
+        // Compile JAVA files with javac, in the "compile.jdk".
+        List<String> command = new ArrayList<>();
+
+        command.add("%s/bin/javac".formatted(System.getProperty("compile.jdk")));
+        command.add("-classpath");
+        command.add(getClassPathOfCompiledClasses());
+        command.add("-d");
+        command.add(classesDir.toString());
+        for (Path path : paths) {
+            command.add(path.toAbsolutePath().toString());
+        }
+
+        executeCompileCommand(command);
+    }
+
+    private List<Path> writeSourcesToFile(List<SourceCode> sources) {
+        List<Path> storedFiles = new ArrayList<Path>();
+        for (SourceCode sourceCode : sources) {
+            Path path = sourceDir.resolve(sourceCode.filePathName());
+            writeCodeToFile(sourceCode.code, path);
+            storedFiles.add(path);
+        }
+        return storedFiles;
+    }
+
+    private static void writeCodeToFile(String code, Path path) {
+        println("File: " + path.toString());
+
+        // Ensure directory of the file exists.
+        Path dir = path.getParent();
+        try {
+            Files.createDirectories(dir);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not create directory: " + dir.toString(), e);
+        }
```

Comment:
Could be extracted to a method `ensureDirectoryExists()`. Same below (i.e. `writeToFile()`).

Reply:
I think it is still small enough.... so I'd rather keep it like it is. Splitting it all up creates a lot of boiler plate.